gooey/gui/util/time.py

```diff
@@ -0,0 +1,93 @@
+"""
+Module for evaluating time elapsed & time remaining from progress
+"""
+import wx
+from gooey.gui.pubsub import pub
+from gooey.gui import events
+
+class Timing(object):
+
+    def __init__(self, parent):
+        self.startTime = 0
+        self.estimatedRemaining = None
+        self.wxTimer = wx.Timer(parent)
+        self.parent = parent
+        parent.Bind(wx.EVT_TIMER, self.publishTime, self.wxTimer)
+
+        pub.subscribe(events.PROGRESS_UPDATE, self._updateEstimate)
```

Comment:
There's a tiny part of me that wants to squabble over this line and as I endlessly bike shed over what goes where, but... I'll spare you my neurotic tendencies ^_^ Changes look good to me. Branch just needs to be brought up to date with current release, then I'll do some final sanity testing on my end, then should be good to go! 

Thanks for all your work on this!

Reply:
I've rebased on the `1.0.5-release` branch, and it's telling me there will be 71 pushes and 38 pulls, unsure if this will become apart of this pull request, and I've changed the target branch to merge into as `1.0.5-release`, let me know if this isn't correct

Furthermore, before I push all these changes onto this, I ran the tests (and tried to run a Gooey application) and I'm getting 65 failures and 1 error (and the application won't run), which all the errors are `wx._core.wxAssertionError: C++ assertion "Assert failure" failed at ..\..\src\msw\font.cpp(586) in wxNativeFontInfo::SetWeight(): unknown font weight`. This seems to be stemming from the console text object, so unsure if I need to alter anything with these changes?

```
Traceback (most recent call last):
  File "c:\Users\jackm\Documents\GitHub\Gooey\gooey\tests\test_time_remaining.py", line 31, in test_time_remaining_visibility_on_complete
    with instrumentGooey(self.make_parser(), timing_options=testdata) as (app, gooeyApp):
  File "c:\users\jackm\.pyenv\pyenv-win\versions\3.8.1-amd64\lib\contextlib.py", line 113, in __enter__
    return next(self.gen)
  File "c:\Users\jackm\Documents\GitHub\Gooey\gooey\tests\harness.py", line 18, in instrumentGooey
    app, gooey = application.build_app(buildspec)
  File "c:\Users\jackm\Documents\GitHub\Gooey\gooey\gui\application.py", line 32, in build_app
    gapp = GooeyApplication(merge(build_spec, imagesPaths))
  File "c:\Users\jackm\Documents\GitHub\Gooey\gooey\gui\containers\application.py", line 47, in __init__
    self.console = Console(self, buildSpec)
  File "c:\Users\jackm\Documents\GitHub\Gooey\gooey\gui\components\console.py", line 27, in __init__
    self.textbox.SetFont(wx.Font(
wx._core.wxAssertionError: C++ assertion "Assert failure" failed at ..\..\src\msw\font.cpp(586) in wxNativeFontInfo::SetWeight(): unknown font weight
```

Reply:
I believe my drama from above was from not updating WxPython from the `1.0.5-release` branch, I switched branches and tried running the tests there, that worked, switched back and everything is hunky dory. 😄

Reply:
I hope I've rebased this right haha 😄 

Let me know if any dramas

Reply:
Hmm.. yeah, looks like there may have been some drama in that rebase ^_^ From looking at the commits, I'm guessing you rebased in the wrong 'direction'. 1.0.5 got layered on top of your commits, thus mangling history (and thoroughly confusing Github, which now shows unrelated comments from other entirely different PRs!). 

Easiest way to fix this is probably to grab a fresh copy of 1.0.5, checkout your branch as it existed before you did the rebase, and then do the rebase again. 

If you don't have your changes handy on a backup branch, you can get back to your original state by using git's reflog. 

```
git reflog
```
Here's an example output. This is me rebasing your branch locally on my machine. The bottom sha is the the one before any rebasing took place. 
```
8cdd582 HEAD@{39}: rebase: Add documentation about clear_before_run
8cdbbfe HEAD@{40}: rebase: Add Listbox height option to documentation
b93f4d5 HEAD@{41}: rebase: closes #625 - terminal font and colors not being set correctly
bdbc6c8 HEAD@{42}: rebase: Update french.json
ef3b011 HEAD@{43}: rebase: Fix typo in ISSUE_TEMPLATE.md.
5bb3355 HEAD@{44}: rebase: Ensure OnDropFiles function matches interface
da33e37 HEAD@{45}: rebase: allow control+scrollwheel to zoom rich text console #432
620f797 HEAD@{46}: rebase: add missing return statement in processor.py
e554e38 HEAD@{47}: rebase: closes #592 - clicking close button can cause Gooey to hang
f7ae29a HEAD@{48}: rebase: Do not translate arguments in add_subparsers
833d01e HEAD@{49}: rebase: add test cases for show_stop_warning decorator option
45f4753 HEAD@{50}: rebase: Fix show_stop_warning
1619caa HEAD@{51}: rebase: Resolve all wx warnings and errors and bump version to 4.1.0
285e75f HEAD@{52}: rebase: Added test for parent inheritance behaviour and made "widgets" behave the same as "defaults" for duplicate named arguments.
3bd7e9b HEAD@{53}: rebase: Copy widgets and their configuration from parent parsers. Resolves #608
4e4a661 HEAD@{54}: rebase: Prevent crash when getting dynamic properties
264561f HEAD@{55}: rebase: Add missing chinese translations
7ae42ac HEAD@{56}: rebase: excess space
2f3aeda HEAD@{57}: rebase: typo
78181c2 HEAD@{58}: rebase: closes #593 & #591 - dynamic updates causes slowdowns and loses selection
4cded5d (origin/master, master) HEAD@{59}: rebase: checkout master
29e8fc0 (JackMcKew-1.0.4-release-tqdm-loading-bar) HEAD@{60}: checkout: moving from 1.0.5-release to 1.0.5-release-bk
``` 

You can create a new branch from that pre-rebase sha and be right back where you started

```
git checkout -b you-new-branch-name relevant-sha-from-reflog
```

I suspect 1.0.5 is mangled, so I'd say blow it away locally and then grab it from origin 

```
git branch -D 1.0.5-release 
git fetch 
git checkout -b 1.0.5-release origin/1.0.5-release 
```
That'll put the latest clean version on your machine and swap you to it.  

Finally, checkout your dev branch again 
```
git checkout the-branch-you-created-above-from-the reflog 
```

then rebase from_this_ branch by pointing at 1.0.5-release 

```
git rebase 1.0.5-release 
```

Hopefully that helps!

Reply:
Thank you so much for the detailed guide, it helped a lot!

It seemed to be quite a painful git experience but I think we're there, what I had to do was:

1. Create a new branch locally of this PR
2. Hard reset to this sha `git reset --hard 2e56d21`
3. Rebase to `1.0.5-release`
4. Do a force push to push the new local branch onto this PR remote branch `git push -f origin 1.0.4-release-tqdm-loading-bar-rebase:1.0.4-release-tqdm-loading-bar`

All my local testing works, the tests pass, and the PR is only showing 1 commit so I think this is the winner! 😄

Reply:
Commits look good! `git` can definitely be a bit of a beast to wrangle. 

Just a few remaining minor issues which popped up during local testing. Should be quick fixes. 

1. The progress bar is displaying on the config screen when you launch Gooey: ![image](https://user-images.githubusercontent.com/1408720/95670575-03892a00-0b42-11eb-8cad-a7b52c118e16.png)

2. If you run with the timing options enabled but _don't_ supply `progress` args, you get errors in `timing.py`

```python
@Gooey(
        program_name="Elapsed / Remaining Timer on Progress in Gooey",
        # progress_regex=r"^progress: (?P<current>\d+)/(?P<total>\d+)$",
        # progress_expr="current / total * 100",
        timing_options = {
                'show_time_remaining':False,
                'show_time_remaining_on_complete':False
        }
        )
```

Specifically [this line](https://github.com/chriskiehl/Gooey/pull/589/files#diff-be5e5016522e8f934cb056dffbf2908aR21) here:
```python
if(prog > 0)
```

Now, obviously, under ideal circumstances, people wouldn't turn on the time estimation feature without also supplying a way for the estimates to be _done_. However, the API allows it, which means we should gracefully handle it and not throw errors. 

Longer term, I'll be tightening up the API validation for all of Gooey's arguments (such that we can easily say A depends on B). But for now, you'll have to just patch around the lack of that capability so that your code doesn't throw errors if the `progress` args haven't been supplied.

Reply:
These are both now fixed and I even tidied up the footer label to be a bit cleaner 😄